### DISCUSSION TITLE
Ditch ADB for custom Rust implementation

### DISCUSSION THREAD

**Discussion by lavafroth** [Universal-Debloater-Alliance/universal-android-debloater-next-generation#643] [Show and tell]
Thanks to the commendable efforts of @cocool97, we can interact with android devices through a custom Rust implementation of ADB. No binary needed! Check out the project [here](https://github.com/cocool97/adb_client/), specifically the [USB branch](https://github.com/cocool97/adb_client/tree/usb). It is still in the works and does not have the `adb shell` command implemented yet but you can try it out from [my branch](https://github.com/lavafroth/adb_client/tree/shell)

Here's a screenshot:
![Screenshot_20240923_132435](https://github.com/user-attachments/assets/37df41e3-34fc-4553-8206-d9fe7826ed29)

**Comment by cocool97** [comment]
Thanks for pointed me ! 

As said work on USB is still under progress, we are implementing it step by step. Currently everything (like shell, push, pull...) can still be done using an intermediate ADB server, but everything will also be implemented with USB directly, removing this unneeded dependency.

**Comment by AnonymousWP** [comment]
So once that's done, I will convert this to an issue.

**Comment by Rudxain** [comment]
I mentioned it [here](https://github.com/Universal-Debloater-Alliance/universal-android-debloater-next-generation/issues/508#issuecomment-2366911021)
> it would allow:
> - us to choose an exact ADB version (more stable and predictable than an auto-updatable ADB)
> - the compiler to further optimize and shrink the sizes of UAD+ADB

I was considering that we should contribute to that project, as that would benefit the whole community! (not only UAD)

**Reply by AnonymousWP** [reply]
> I was considering that we should contribute to that project, as that would benefit the whole community! (not only UAD)

Even though I agree with you, I said it before: we don't even have enough developers/activity for active development (it usually comes and it goes, which is fine as it's voluntary), so maintaining this project is already a challenge, let alone also contributing to other repos. And it's not really a priority, as UAD-ng does what it needs to do right now, in essence.

**Comment by cocool97** [comment]
Hi everyone ! 

USB support for `push`, `pull`, `shell`, `stat` and `reboot` commands has been added to `adb_client` crate's today release (version 2.0.0)  ! You can try it if you want, feedbacks are always welcome :)

Thanks to @lavafroth from this community that has implemented `pull` and `stat` commands and some useful functions !

**Reply by Rudxain** [reply]
Awesome! Props to @lavafroth !

I've opened an issue about this: #700
